File: airtools/logmart.py
#!/usr/bin/env python
"""
solve b=Ax using parallel log-entropy MART  (De Pierro 1991)

Original Matlab by Joshua Semeter
port to Python by Michael Hirsch
"""
import numpy as np
import math


def logmart(A: np.ndarray, b: np.ndarray,
            *,
            relax: float = 1.,
            x0: float = None,
            sigma: float = 1.,
            max_iter: int = 20) -> tuple:
    """
    estimation halted based on chi**2 value
    A and b must be all NON-NEGATIVE!

    Parameters
    ----------
    A: numpy.ndarray
        NxM array "projection"
    b: numpy.ndarray
        N column vector "observation"

    Returns
    -------
    x_est: numpy.ndarray
        M column vector estimate of "true" x in b = A @ x

    Matlab logmart.m AUTHOR: Joshua Semeter 5-2015

    >>> A = np.diag([5, 5, 5])
    >>> x = np.array([1,2,3])
    >>> b = A @ x
    """
# %% parameter check
    if b.ndim != 1:
        raise ValueError('y must be a column vector')
    if A.ndim != 2:
        raise ValueError('A must be a matrix')
    if A.shape[0] != b.size:
        raise ValueError('A and b: number of rows must match')
    if not isinstance(relax, (int, float)):
        raise ValueError('relax is a scalar float')
    if (A < 0).any():
        raise ValueError('A must be all non-negative')
    if (b < 0).any():
        raise ValueError('b must be all non-negative')

    b = b.copy()  # needed to avoid modifying outside this function!
    # %% make sure there are no 0's in b
    b[b <= 1e-8] = 1e-8
# %% set defaults
    if x0 is None:  # backproject
        x = A.T @ b / A.sum()
        x *= b.max() / (A @ x).max()
    elif isinstance(x0, (float, int)) or x0.size == 1:  # replicate
        x = x0 * np.ones_like(b)
    else:
        x = x0
    if not x.size == A.shape[1]:
        raise ValueError('x0 must be scalar or match Ncolumns of A')

    x[x < 1e-8] = 1e-8
    # W=sigma;
    # W=linspace(1,0,size(A,1))';
    # W=rand(size(A,1),1);
    W = np.ones(A.shape[0])
    W = W / W.sum()

    chi2 = chi_squared(A, b, x, sigma)
# %%  iterate solution, plot estimated data (diag elems of x#A)
    for i in range(max_iter):
        x_prev = x
        xA = A @ x
        t = (1/xA).min()
        C = relax * t * (1 - xA/b)
        x /= (1 - x*(A.T @ (W*C)))
# %% monitor solution
        chiold = chi2
        chi2 = chi_squared(A, b, x, sigma)
        if i > 1 and chi2 >= chiold:
            break
        # if chi2 < 0.7:
        #    break

    return x_prev, chi2, i


def chi_squared(A: np.ndarray, b: np.ndarray, x: np.ndarray, sigma: float) -> float:
    return math.sqrt((((A @ x - b)/sigma)**2).sum())

File: +airtools/logmart.m
%solve y=Ax using parallel log-entropy MART  (De Pierro 1991)
function [ x,y_est,chi2,i ] = logmart( y,A,relax,x0,sigma,max_iter )
% Program is stopped if Chisquare increases.
% A is NxM array
% Y is Nx1 vector
% returns Mx1 vector
%
% KEYWORDS:
% relax		user specified relaxation constant	(default is 20.)
% x0	user specified initial guess (N vector)
%           (default is backproject y, i.e., y#A)
% max_iter	user specified max number of iterations (default is 20)
%
%AUTHOR:	Joshua Semeter
%LAST MODIFIED:	5-2015

%  Simple test problem
% A=diag([5 5 5]);
% x=[1;2;3];
% y=A*x;
arguments
y (:,1) {mustBeNumeric,mustBeNonnegative}
A (:,:) {mustBeNumeric,mustBeNonnegative}
relax (1,1) {mustBePositive} = 20
x0 {mustBeNumeric,mustBeNonnegative} = []
sigma (1,1) {mustBeNumeric,mustBePositive} = 1
max_iter (1,1) {mustBeInteger, mustBePositive} = 20
end
%% make sure there are no 0's in y
y(y<=1e-8) = 1e-8;

if isempty(x0)
    x=(A'*y)./sum(A(:));
    xA=A*x;
    x=x.*max(y(:))/max(xA(:));
else
    x=x0;
end

% W=sigma;
% W=linspace(1,0,size(A,1))';
% W=rand(size(A,1),1);
W=ones(size(A,1),1);
W=W./sum(W);

chi2 = chi_squared(y, A, x, sigma);

for i = 1:max_iter
%%  iterate solution, plot estimated data (diag elems of x#A)
  xold=x;
  t=min(1./(A*x));
  C=(relax*t*( 1-((A*x)./y) ));
  x=x./(1-x.*(A'*(W.*C)));
%% monitor solution
  chiold=chi2;
  chi2 = chi_squared(y, A, x, sigma);
%   dchi2=(chi2-chiold);
  if ((chi2>chiold) && (i>2)), break, end  %  || (chi2<0.7)

%figure(9); clf; hold off;
%Nest=reshape(x,69,83);
%imagesc(Nest); caxis([0,1e11]);
%set(gca,'YDir','normal'); set(gca,'XDir','normal');
%pause;

end % for
x=xold;
if nargout>1
    y_est=(A*x);
end
end %function

function chi2 = chi_squared(y, A, x, sigma)
chi2 = sqrt( sum(((A*x - y)./sigma).^2) );
end
